runtime/compiler/aarch64/codegen/CallSnippet.cpp

```diff
@@ -0,0 +1,590 @@
+
+#include "codegen/CallSnippet.hpp"
+#include "codegen/ARM64AOTRelocation.hpp"
+#include "codegen/CodeGenerator.hpp"
+#include "codegen/InstOpCode.hpp"
+#include "codegen/Linkage.hpp"
+#include "codegen/Machine.hpp"
+#include "codegen/Register.hpp"
+#include "env/CompilerEnv.hpp"
+#include "env/J2IThunk.hpp"
+#include "env/VMJ9.h"
+#include "il/Node.hpp"
+#include "il/Node_inlines.hpp"
+#include "il/symbol/LabelSymbol.hpp"
+#include "runtime/CodeCacheManager.hpp"
+
+static uint8_t *storeArgumentItem(TR::InstOpCode::Mnemonic op, uint8_t *buffer, TR::RealRegister *reg, int32_t offset, TR::CodeGenerator *cg)
+   {
+   TR::RealRegister *stackPtr = cg->getLinkage()->cg()->getStackPointerRegister();
+   TR::InstOpCode opCode(op);
+   uint32_t enc = (uint32_t)opCode.getOpCodeBinaryEncoding();
+   TR_ASSERT((enc & 0x3b200000) == 0x39000000, "Instruction not supported in storeArgumentItem()");
+
+   uint32_t size = (enc >> 30) & 3; /* b=0, h=1, w=2, x=3 */
+   uint32_t shifted = offset >> size;
+   uint32_t *wcursor = (uint32_t *)buffer;
+
+   opCode.copyBinaryToBuffer(buffer);
+   reg->setRegisterFieldRT(wcursor);
+   stackPtr->setRegisterFieldRN(wcursor);
+   *wcursor |= (shifted & 0xfff) << 10; /* imm12 */
+
+   return buffer+4;
+   }
+
+static uint8_t *flushArgumentsToStack(uint8_t *buffer, TR::Node *callNode, int32_t argSize, TR::CodeGenerator *cg)
+   {
+   uint32_t intArgNum=0, floatArgNum=0, offset;
+   TR::Machine *machine = cg->machine();
+   const TR::ARM64LinkageProperties &linkage = cg->getLinkage(callNode->getSymbol()->castToMethodSymbol()->getLinkageConvention())->getProperties();
+   int32_t argStart = callNode->getFirstArgumentIndex();
+
+   if (linkage.getRightToLeft())
+      offset = linkage.getOffsetToFirstParm();
+   else
+      offset = argSize+linkage.getOffsetToFirstParm();
+
+   for (int32_t i=argStart; i<callNode->getNumChildren();i++)
+      {
+      TR::Node *child = callNode->getChild(i);
+      switch (child->getDataType())
+         {
+         case TR::Int8:
+         case TR::Int16:
+         case TR::Int32:
+            if (!linkage.getRightToLeft())
+               offset -= 4;
+            if (intArgNum < linkage.getNumIntArgRegs())
+               {
+               buffer = storeArgumentItem(TR::InstOpCode::strimmw, buffer, machine->getRealRegister(linkage.getIntegerArgumentRegister(intArgNum)), offset, cg);
+               }
+            intArgNum++;
+            if (linkage.getRightToLeft())
+               offset += 4;
+            break;
+         case TR::Int64:
+         case TR::Address:
+            if (!linkage.getRightToLeft())
+               offset -= 8;
+            if (intArgNum < linkage.getNumIntArgRegs())
+               {
+               buffer = storeArgumentItem(TR::InstOpCode::strimmx, buffer, machine->getRealRegister(linkage.getIntegerArgumentRegister(intArgNum)), offset, cg);
+               }
+            intArgNum += 2;
```

Comment:
Why +2?

Reply:
I just overlooked this when I wrote the code based on the `arm` version.